test/hw8/test_task2.py

```diff
@@ -0,0 +1,114 @@
+import os
+import sqlite3
+import tempfile
+import uuid
+
+import pytest
+
+from All_home_works.hw8.task2 import TableData
+
+
+@pytest.fixture(scope="module")
+def create_db_table():
+    directory = tempfile.TemporaryDirectory()
+    path = os.path.join(directory.name, "test.sqlite")
+    db_name = os.path.normpath(path)
+    db_table = f"test_table_{uuid.uuid4()}"
+
+    conn = sqlite3.connect(db_name)
+    cursor = conn.cursor()
+    cursor.execute(f"CREATE TABLE '{db_table}' (name, author)")  # noqa: S608
+    cursor.execute(
+        f"INSERT INTO '{db_table}' (name, author) VALUES ('Book1', 'John Doe')"  # noqa: S608
+    )
+    cursor.execute(
+        f"INSERT INTO '{db_table}' (name, author) VALUES ('Book2', 'Jain Doe')"  # noqa: S608
+    )
+    conn.commit()
+    yield db_name, db_table
+    cursor.execute(f"DROP TABLE '{db_table}'")  # noqa: S608
+    conn.commit()
+    conn.close()
+    directory.cleanup()
+
+
+@pytest.fixture()
+def updated_db(create_db_table):
+    db_name, db_table = create_db_table
+    storage = TableData(db_name, db_table)
+
+    conn = sqlite3.connect(db_name)
+    cursor = conn.cursor()
+    cursor.execute(
+        f"INSERT INTO '{db_table}' (name, author) VALUES ('Test Book', 'Test Author')"  # noqa: S608
+    )
+    conn.commit()
+    yield storage
+    cursor.execute(f"DELETE FROM '{db_table}' WHERE name='Test Book'")  # noqa: S608
+    conn.commit()
+    conn.close()
+
+
+def test_table_data__if_len_correct(create_db_table):
+    db_name, db_table = create_db_table
+    storage = TableData(db_name, db_table)
+    assert 2 == len(storage)
+
+
+def test_table_data__if_len_correct__after_update_storage(updated_db):
+    storage = updated_db
+    assert 3 == len(storage)
+
+
+def test_table_data__get_row_from_db_by_name(create_db_table):
+    db_name, db_table = create_db_table
+    storage = TableData(db_name, db_table)
+    assert {"name": "Book1", "author": "John Doe"} == storage["Book1"]
+    assert None is storage["Book3"]
+
+
+def test_table_data__get_row_from_db_by_name__check_storage_after_update(updated_db):
+    storage = updated_db
+    assert {"name": "Test Book", "author": "Test Author"} == storage["Test Book"]
+
+
+def test_table_data__for_loop(create_db_table):
+    db_name, db_table = create_db_table
+    storage = TableData(db_name, db_table)
+    iter_storage = iter(storage)
+    row = next(iter_storage)
```

Comment:
```
>>> x =  iter(storage)
>>> x
x
>>> list(x) 
[x, x]
```

I found it extremely confusing, that iterator returns instances of itself during iteration.
I expect that the iterator returns a new value each time.  Instead of that, it returns the same value and mutates it.

Please avoid any mutations after object was intialized.

Reply:
> if iter will return a dict or a list: 

What should be returned by `__iter__` according to the documentation?

How `in` and `__iter__` work together?

Reply:
`__in__`
`a __in__ (c, b, a):`
   `a __eq__ c`
   `a __eq__ b`
   `a __eq__ a`

`__iter__` in my case returns:
`(x, x, x)`
  `__in__` in my case:
  `a __eq__ x`
     `__eq__` in my case:
     `a in x.__dict__.values()`

`__iter__` if it will return a list:
([name1, age1], [name2, age2])
   `__in__` in my case:
  `name1 __eq__ [name1, age1] `
Or did I something miss?

Reply:
>What should be returned by __iter__ according to the documentation?

> object implements iteration protocol. i.e. you could use it in for loops::
for president in presidents:
print(president['name'])

There is no info about what should returns in `__iter__` except it has `__getitem__`

Reply:
> object implements iteration protocol. i.e. you could use it in for loops::

Not only iterators could be used in the for loops. 

> There is no info about what should returns in __iter__ 

Try to dig into https://docs.python.org

Reply:
I am not sure if I understand you correctly
```
__eq__ in my case:
a in x.__dict__.values()
```


```
"a" == {1: "a", 2: "b"} # True
"b" == {1: "a", 2: "b"} # True
```

Than I hear equals I expect transitivity that means  "a" == "b" should be True as well;

Reply:
In general, I want to say that I do not know how to make a dictionary return in `for loop`, and when we do
`something IN storage`, it compared not with the dictionary, but with the values of the dictionary.
Because when we do it: `something IN storage` it do this: 
`for i in storage`
`if something == i`
`return True`

I mean it also calls `__iter__` and compares `something` with `__iter__` returns

Reply:
I got your point. 

But lets me rewrite this code in different way: 

```python
i = next(storage)
while True:
    next(storage)
    if something == i
        return True
```

This is usly and won't work with most implementation of iterator protocall, except this.